src/linalg.jl

```diff
@@ -629,6 +629,61 @@ function dot(A::AbstractSparseMatrixCSC, B::Union{DenseMatrixUnion,WrapperMatrix
     return conj(dot(B, A))
 end
 
+function dot(x::AbstractSparseVector, Q::Diagonal, y::AbstractVector)
+    if length(x) != length(y)
```

Comment:
I think you also need to check the length of `Q.diag`. Perhaps you could replace `Q` with `D` and continue to use `d = D.diag`?

Reply:
both done thank you!